File: Chapter 02/ObserverDesignPatternExample.py
class ObservableSubject:
    def __init__(self):
        self._observers = []
    def register_observer(self, observer):
        self._observers.append(observer)
    def remove_observer(self, observer):
        self._observers.remove(observer)
    def notify_observers(self):
        for observer in self._observers:
            observer.update(self)

class WeatherChannel(ObservableSubject):
    def __init__(self):
        super().__init__()
        self._temperature = 0
    def set_temperature(self, temperature):
        self._temperature = temperature
        self.notify_observers()
    def get_temperature(self):
        return self._temperature

class Observer:
    def update(self, subject):
        pass

class TemperatureDisplay(Observer):
    def update(self, subject):
        if isinstance(subject, WeatherChannel):
            temperature = subject.get_temperature()
            print(f"Temperature: {temperature}°C")

if __name__ == "__main__":
    weather_channel = WeatherChannel()
    display1 = TemperatureDisplay()
    display2 = TemperatureDisplay()
    weather_channel.register_observer(display1)
    weather_channel.register_observer(display2)
    weather_channel.set_temperature(25)

File: Chapter 02/ObserverDesignPatternExample.java
import java.util.ArrayList;
import java.util.List;

interface ObservableSubject {
    void registerObserver(Observer observer);
    void removeObserver(Observer observer);
    void notifyObservers();
}

class WeatherChannel implements ObservableSubject {
    private int temperature;
    private List<Observer> observers = new ArrayList<>();

    public void setTemperature(int temperature) {
        this.temperature = temperature;
        notifyObservers();
    }

    @Override
    public void registerObserver(Observer observer) {
        observers.add(observer);
    }

    @Override
    public void removeObserver(Observer observer) {
        observers.remove(observer);
    }

    @Override
    public void notifyObservers() {
        for (Observer observer : observers) {
            observer.update(temperature);
        }
    }
}

interface Observer {
    void update(int temperature);
}

class TemperatureDisplay implements Observer {
    private int temperature;

    @Override
    public void update(int temperature) {
        this.temperature = temperature;
        display();
    }

    public void display() {
        System.out.println("Temperature: " + temperature + "°C");
    }
}
public class ObserverDesignPatternExample {
    public static void main(String[] args) {
        WeatherChannel weatherChannel = new WeatherChannel();
        TemperatureDisplay display1 = new TemperatureDisplay();
        TemperatureDisplay display2 = new TemperatureDisplay();

        weatherChannel.registerObserver(display1);
        weatherChannel.registerObserver(display2);
        weatherChannel.setTemperature(25);
    }
}
